here-naksha-cli/src/jvmMain/java/com/here/naksha/cli/copy/StorageConfigConverter.java

```
@@ -0,0 +1,42 @@
+package com.here.naksha.cli.copy;
+
+import naksha.base.JvmBoxingUtil;
+import naksha.base.Platform;
+import naksha.model.objects.NakshaStorage;
+import picocli.CommandLine;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+class StorageConfigConverter implements CommandLine.ITypeConverter<File> {
+    public File convert(String value) throws CommandLine.TypeConversionException {
+        File file = new File(value);
+        Path path = file.toPath(); // if needed
+
+        if(!file.exists()) {
+            throw new CommandLine.TypeConversionException("File does not exist!");
+        }
+
+        if(!file.isFile()) {
+            throw new CommandLine.TypeConversionException("It is not a file!");
+        }
+
+        String json;
+
+        try {
+            json = Files.readString(path);
+        } catch (Exception e) {
+            throw new CommandLine.TypeConversionException("Problem with reading!");
+        }
+
+        try {
+            Object rawConfig = Platform.fromJSON(json);
+            JvmBoxingUtil.box(rawConfig, NakshaStorage.class);
+        } catch (Exception e) {
+            throw new CommandLine.TypeConversionException("Problem with json parsing!");
```

I have two issues with this `catch` clause:
1) If you are throwing exception that came to life because of other exception - include the root cause. Most exception classes accept `Exception` instance as last constructor parameter, effectively setting `cause` under the hood. This would enrich the stacktrace - so that when we have a failure, we get more data on why it happened.
2) Is `CommandLine.TypeConversionException` the best exception here? We don't convert parameter type here, we are already past this stage. The potential exception is because of `Platform.fromJSON` or `JvmBoxingUtil.box`

---

Actually, the remark  no 2 is for all exceptions thrown in this class - we always throw ` CommandLine.TypeConversionException` 
Take a look at the description of this class:
```
Exception thrown by CommandLine. ITypeConverter implementations to indicate a String could not be converted.
```
This is about string parsing, and we throw it for:
- non existing file
- invalid file
- invalid naksha deserialization

IMO there are better exceptions for all of these cases

---

The picocli provides only these exceptions (without the CommandLine instance as a parameter):
- TypeConversionException, which I used
- and top level PicocliException

The interface which I implemented requires TypeConversionException to have more friendly message to the user.

```
    public interface ITypeConverter<K> {
        /**
         * Converts the specified command line argument value to some domain object.
         * @param value the command line argument String value
         * @return the resulting domain object
         * @throws Exception an exception detailing what went wrong during the conversion.
         *          Any exception thrown from this method will be caught and shown to the end user.
         *          An example error message shown to the end user could look something like this:
         *          {@code Invalid value for option '--some-option': cannot convert 'xxxinvalidinput' to SomeType (java.lang.IllegalArgumentException: Invalid format: must be 'x:y:z' but was 'xxxinvalidinput')}
         * @throws TypeConversionException throw this exception to have more control over the error
         *          message that is shown to the end user when type conversion fails.
         *          An example message shown to the user could look like this:
         *          {@code Invalid value for option '--some-option': Invalid format: must be 'x:y:z' but was 'xxxinvalidinput'}
         */
        K convert(String value) throws Exception;
    }
```

What is your suggestion?

---

And the TypeConversionException does not accept root cause.

---

First of all - it's not true that "The interface which I implemented requires TypeConversionException" - you can throw any exception, they have included one in the doc.

Well, there are three options:
1) Leave it as is. Pros: we follow their recommendation. Cons: we miss on details.
2) Throw exception that is not `TypeConversionException`, just general one, like `IllegalArgumentException` - then see how does it look. If it's not messy, I'd go with that.
3) We could experiment and create a subtype of `TypeConversionException` - the constructor would accept both `msg` and `cause` - at first you call `super(msg)` and then call `initCause(cause)`.

```
class CustomException extends CommandLine.TypeConversionException {
                                                                          
    public CustomException(String msg) {
        super(msg);
    }
                                                                          
    public CustomException(String msg, Throwable cause) {
        super(msg);
        initCause(cause);
    }
}
```

---

For now - I am ok with any of these approaches :) Maybe try each of them, make the command fail and see for yourself which looks the best? Especially for cases that are not super trivial (ie error thrown in `box` method)